lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,354 @@
+        add_auto(post, options)
+        add_card_on_file(post, options)
+        add_hotel(post, options)
+        add_pin(post, options)
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if credit_card.respond_to?(:number)
+          post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+          post[:card][:number] = credit_card.number
+          post[:card][:entryMode] = options[:entry_mode]
+          post[:card][:present] = options[:present]
+        else
+          post[:card] = {} if post[:card].nil?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = credit_card
+        end
+        if options[:card].present?
+          post[:card][:expirationDate] = options[:card][:expiration_date] if options[:card][:expiration_date]
+          post[:card][:entryMode] = options[:card][:entry_mode]
+          post[:card][:present] = options[:card][:present]
+          post[:card][:securityCode] = {} unless post[:card][:securityCode]
+          post[:card][:securityCode][:indicator] = options[:card][:indicator] if options[:card][:indicator]
+          post[:card][:securityCode][:value] = options[:card][:value] if options[:card][:value]
+          post[:card][:securityCode] = {}
+        end
+        if options[:security_code].present?
+          post[:card][:securityCode] = {}
+          post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+          post[:card][:securityCode][:value] = options[:security_code][:value]
+        end
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
```

Comment:
`address` data will consistently be presented in the `billing_address` sub-hash of the `options` hash. I'll leave another comment in the remote test file to suggest a way to build it to fit the common pattern in AM.